### PR TITLE
[setup] drop deprecated `distutils` usage

### PR DESCRIPTION
**Pull request by XuehaiPan**
# What does this PR do?

Migrate setup script to `pyproject.toml` ([PEP 517 – A build-system independent format for source trees](https://peps.python.org/pep-0517)).

Changes:

- drop deprecated `distutils` usage
- ~~migrate setup script to `pyproject.toml`~~
- ~~migrate `isort` and `pytest` configs to `pyproject.toml`~~
- ~~migrate `flake8` configs to `.flake8` and remove `setup.cfg` file~~

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Review by sgugger — COMMENTED** [review]
Thanks for your PR, but we are very happy with the setup as it is.

**Comment by XuehaiPan** [comment]
> but we are very happy with the setup as it is.

Thanks for the clarification. The `pyprojct.toml` format is recommended packaging method in [PEP 517 – A build-system independent format for source trees](https://peps.python.org/pep-0517). I have reverted some of the commits but kept the first one. The `distutils` module is deprecated and will be removed in Python 3.12. (See also [PEP 632 – Deprecate distutils module](https://peps.python.org/pep-0632))

File: setup.py [hunk 1]
```diff
@@ -76,24 +76,29 @@
 import os
 import re
 import shutil
-from distutils.core import Command
+import textwrap
```
Comment (sgugger) [inline setup.py:79]: No need to add this.

File: setup.py [hunk 2]
```diff
@@ -76,24 +76,29 @@
...
+            Warning: {stale_egg_info} exists.
+
+            If you recently updated transformers to 3.0 or later, this is expected,
+            but it may prevent transformers from installing in editable mode.
+
+            This directory is automatically generated by Python's packaging tools.
+            I will remove it now.
+
+            See https://github.com/pypa/pip/issues/5466 for details.
+            """
```
Comment (sgugger) [inline setup.py:100]: We can just leave the f-string. The likelihood of this warning popping is next to 0 now that version 2.xxx are so old.

File: setup.py [hunk 3]
```diff
@@ -76,24 +76,29 @@
 import os
 import re
 import shutil
-from distutils.core import Command
+import textwrap
 from pathlib import Path
 
-from setuptools import find_packages, setup
+from setuptools import find_packages, setup, Command
```
[the remark this answers is not in the document]
Comment (XuehaiPan) [reply setup.py:81]: Yes.

**Review by sgugger — APPROVED** [review]
Thanks for your contribution!

File: setup.py (hunk 2 above)
Comment (XuehaiPan) [reply setup.py]: I do not change the string output but refactored it with `textwrap` without `\n`s. I have reverted it if is not relevant to this PR.

**Comment by sgugger** [comment]
I think you can suggest the changes thate removed the setup.cfg in a separate PR, it's a good cleanup (but not relevant to this PR anymore)

For migrating the setup.py to the pyproject, let's see with @LysandreJik what he thinks. My first reaction is to keep what's been working for us all these years ;-)

**Comment by XuehaiPan** [comment]
Thanks for the clarification. Since we have already dropped Python 3.6 support, `setuptools` works very well with `pyproject.toml` based project. We can move the static parts in `setup.py` to `pyproject.toml`. Note that the optional dependencies are too dynamic, so we still need a `setup.py` file.

Most Python utilities support `pyproject.toml` configuration (`black`, `isort`, `ruff`, `pytest`, ...). And some do not even support other config files like `setup.cfg` (e.g., `black`). I think maintaining configurations in a single file is a good practice.

**Pull request merged**